### PR TITLE
KAFKA-12634 enforce checkpoint after restoration

### PR DESCRIPTION
**Pull request by philipnee**
*We want to ensure checkpointing the progress after completing the restoration to prevent losing the progress and needing to restore from scratch

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 1]
```diff
@@ -255,6 +255,7 @@ public void completeRestoration(final java.util.function.Consumer<Set<TopicParti
                 resetOffsetsIfNeededAndInitializeMetadata(offsetResetter);
                 initializeTopology();
                 processorContext.initialize();
+                maybeCheckpoint(true); // enforce checkpoint upon completing restoration
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: As the ticket says, we can only write this checkpoint if we process under at-least-once (ALOS) processing guarantee , but we must not write the checkpoint when we process under exactly-once (EOS). In EOS, we need to rebuild local state when the Streams client crashed or closed dirty. Otherwise, the Streams would reprocess the  records from the last committed offset and add the those same records again to the local state store violating EOS. We know that a Streams client crashed or closed dirty if there is no checkpoint file in the state directory. Now, if we write a checkpoint right after restoration and the Streams client processes records, but before the first commit the Streams client crashes or closes dirty, it would violate EOS for the aforementioned reasons.   
So you need to differentiate between ALOS and EOS similar when a commit is executed: https://github.com/apache/kafka/blob/98c2f88e1c605195ccfac19c49a83216e26146a1/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java#L492

**Review by cadonna — COMMENTED** [review]
Thanks for the PR @philipnee !

Here my feedback!

**Comment by philipnee** [comment]
There's a bunch of failing DedicatedMirrorIntegrationTest - most likely unrelated. There's also a bunch of MetadataQuorumCommandTest, which also seem unrelated. Rebasing the branch.

**Comment by philipnee** [comment]
@cadonna - Thanks for the feedback there.  I added the EOS check before the checkpoint. I also made a bunch of changes to the tests.  To make things obvious, I added a comment to these altered tests to let people know about the checkpointing.

**Comment by guozhangwang** [comment]
Sorry for jumping into this PR late @philipnee , for this ticket, we have fixed it in the new state updater thread ([KAFKA-10199](https://issues.apache.org/jira/browse/KAFKA-10199). And we are expecting to enable the state updater soon. So I'd suggest we do not try to fix it inside the main stream thread any more. WDYT?

**Comment by philipnee** [comment]
Oh yeah, I was wondering about that, too - it's up to you. We can abandon this and wait for 10199 to be completed.

**Comment by philipnee** [comment]
i took a peek at it because @mjsax mentioned it is easy to work on.

**Comment by cadonna** [comment]
> Sorry for jumping into this PR late @philipnee , for this ticket, we have fixed it in the new state updater thread ([KAFKA-10199](https://issues.apache.org/jira/browse/KAFKA-10199). And we are expecting to enable the state updater soon. So I'd suggest we do not try to fix it inside the main stream thread any more. WDYT?

@guozhangwang 
Do you think this change is too risky? 
I do not know how long "soon" means. At the moment, it seems It will still take us some time until we will enable the state updater by default. Even then, we will keep the option to switch back to the old restoration for a couple of releases. 
So, if we think this PR is not too risky, I would go for it. 
Maybe I am a bit out of the loop currently and I am missing something.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 2]
```diff
@@ -1503,7 +1505,9 @@ public void shouldReadCommittedOffsetAndRethrowTimeoutWhenCompleteRestoration()
     public void shouldReInitializeTopologyWhenResuming() throws IOException {
         stateDirectory = EasyMock.createNiceMock(StateDirectory.class);
         EasyMock.expect(stateDirectory.lock(taskId)).andReturn(true);
+        EasyMock.expect(recordCollector.offsets()).andReturn(Collections.emptyMap()); // restoration checkpoint
         EasyMock.expect(recordCollector.offsets()).andThrow(new AssertionError("Should not try to read offsets")).anyTimes();
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: nit:
```suggestion
        EasyMock.expect(recordCollector.offsets()).andReturn(Collections.emptyMap())
            .andThrow(new AssertionError("Should not try to read offsets")).anyTimes();
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 3]
```diff
@@ -1699,7 +1705,8 @@ public void shouldReturnOffsetsForRepartitionTopicsForPurging() {
         consumer.assign(asList(partition1, repartition));
         consumer.updateBeginningOffsets(mkMap(mkEntry(repartition, 0L)));
 
-        EasyMock.expect(stateManager.changelogPartitions()).andReturn(Collections.emptySet());
+        EasyMock.expect(stateManager.changelogOffsets()).andReturn(Collections.emptyMap()); // restoration checkpoint
+        EasyMock.expect(stateManager.changelogPartitions()).andReturn(Collections.emptySet()).times(2);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Since for this test these calls are just stubs (i.e., nothing needs to be verified about these calls), I would replace them with:
```suggestion
EasyMock.expect(stateManager.changelogOffsets()).andStubReturn(Collections.emptyMap()); // restoration checkpoint
        EasyMock.expect(stateManager.changelogPartitions()).andStubReturn(Collections.emptySet());
```

**Review by cadonna — COMMENTED** [review]
Thank you for the PR, @philipnee !

I am missing two tests, one that tests that a checkpoint is written when restoration is completed under at least once and the other that a checkpoint is not written when restoration is completed under exactly once.
I am aware that this is probably implicitly tested in other tests but I would like to have explicit tests for this two cases.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java (hunk 3 above)
Comment (philipnee) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Thanks, I think that make sense.

**Comment by philipnee** [comment]
Hey @cadonna - Thanks for getting back on this. I added two tests covering EOS enabled/disabled cases - LMK if this is what you are looking for. I tried to make the test case explicit there.

**Comment by philipnee** [comment]
Not entirely sure what this is: `[2023-03-29T05:12:47.884Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-13269/core/src/test/scala/unit/kafka/zk/ZkMigrationClientTest.scala:436:50: value addOne is not a member of scala.collection.mutable.ArrayBuffer[scala.collection.mutable.Buffer[org.apache.kafka.server.common.ApiMessageAndVersion]]`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 4]
```diff
@@ -1546,8 +1551,9 @@ public void shouldNotCheckpointOffsetsAgainOnCommitIfSnapshotNotChangedMuch() {
         stateManager.checkpoint();
         EasyMock.expectLastCall().once();
         EasyMock.expect(stateManager.changelogOffsets())
-            .andReturn(singletonMap(changelogPartition, 10L))
-            .andReturn(singletonMap(changelogPartition, 20L));
+                .andReturn(singletonMap(changelogPartition, 10L)) // restoration checkpoint
+                .andReturn(singletonMap(changelogPartition, 10L))
+                .andReturn(singletonMap(changelogPartition, 20L));
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]:
```suggestion
            .andReturn(singletonMap(changelogPartition, 10L)) // restoration checkpoint
            .andReturn(singletonMap(changelogPartition, 10L))
            .andReturn(singletonMap(changelogPartition, 20L));
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 5]
```diff
@@ -1692,43 +1699,45 @@ public void shouldReturnOffsetsForRepartitionTopicsForPurging() {
         final TopicPartition repartition = new TopicPartition("repartition", 1);
 
         final ProcessorTopology topology = withRepartitionTopics(
-            asList(source1, source2),
-            mkMap(mkEntry(topic1, source1), mkEntry(repartition.topic(), source2)),
-            singleton(repartition.topic())
+                asList(source1, source2),
+                mkMap(mkEntry(topic1, source1), mkEntry(repartition.topic(), source2)),
+                singleton(repartition.topic())
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Could you revert this change, please? An indentation of 4 is fine.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 6]
```diff
@@ -1692,43 +1699,45 @@ public void shouldReturnOffsetsForRepartitionTopicsForPurging() {
...
-            taskId,
-            config,
-            stateManager,
-            streamsMetrics,
-            null
+                taskId,
+                config,
+                stateManager,
+                streamsMetrics,
+                null
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Indentation of 4 is fine. Could you revert this change, please?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 7]
```diff
@@ -1692,43 +1699,45 @@ public void shouldReturnOffsetsForRepartitionTopicsForPurging() {
...
+                consumer,
+                new TopologyConfig(null, config, new Properties()).getTaskConfig(),
+                streamsMetrics,
+                stateDirectory,
+                cache,
+                time,
+                stateManager,
+                recordCollector,
+                context,
+                logContext);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Thank you for removing the superfluous space! However, the indentation of 4 is fine. Could you revert the indentation, please?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 8]
```diff
@@ -1692,43 +1699,45 @@ public void shouldReturnOffsetsForRepartitionTopicsForPurging() {
...
 
         task.initializeIfNeeded();
-        task.completeRestoration(noOpResetter -> { });
+        task.completeRestoration(noOpResetter -> {
+        });
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: This change is not needed for the purpose of this PR.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 9]
```diff
@@ -2484,6 +2505,34 @@ public void shouldUpdateOffsetIfCorruptedRecordFollowsValid() {
         );
     }
 
+    @Test
+    public void testRestoration_CheckpointWrittenWhenEOSDisabled() {
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: We usually use the form `should...` for test names. My proposal here is to use `shouldCheckpointAfterRestorationWhenAtLeastOnceEnabled`.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 10]
```diff
@@ -2484,6 +2505,34 @@ public void shouldUpdateOffsetIfCorruptedRecordFollowsValid() {
...
+        EasyMock.replay(stateManager, recordCollector);
+
+        task = createStatefulTask(createConfig(AT_LEAST_ONCE, "100"), true);
+        task.initializeIfNeeded();
+        task.completeRestoration(noOpResetter -> { });
+        EasyMock.verify(stateManager);
+    }
+
+    @Test
+    public void testRestoration_CheckpointNotWrittenWhenEOSEnabled() {
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: See aboveWe usually use the form `should...` for test names. My proposal here is to use `shouldNotCheckpointAfterRestorationWhenExactlyOnceEnabled`.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 11]
```diff
@@ -2484,6 +2505,34 @@ public void shouldUpdateOffsetIfCorruptedRecordFollowsValid() {
...
+                .andReturn(singletonMap(partition1, 0L)); // restoration checkpoint
+        EasyMock.expect(recordCollector.offsets()).andReturn(Collections.emptyMap());
+        stateManager.checkpoint();
+        EasyMock.expectLastCall().once(); // checkpoint should only be called once
+        EasyMock.replay(stateManager, recordCollector);
+
+        task = createStatefulTask(createConfig(AT_LEAST_ONCE, "100"), true);
+        task.initializeIfNeeded();
+        task.completeRestoration(noOpResetter -> { });
+        EasyMock.verify(stateManager);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Could you please use Mockito instead of EasyMock here? We are migrating to Mockito and we would like to have new tests using Mockito. See `TaskManagerTest` for an example where a test class uses Mockito and EasyMock.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 12]
```diff
@@ -2484,6 +2505,34 @@ public void shouldUpdateOffsetIfCorruptedRecordFollowsValid() {
...
+    public void testRestoration_CheckpointNotWrittenWhenEOSEnabled() {
+        stateManager.checkpoint();
+        EasyMock.expectLastCall().andStubThrow(new AssertionError("should not checkpoint on EOS"));
+        stateManager.changelogOffsets();
+        EasyMock.expectLastCall().andStubThrow(new AssertionError("should not checkpoint on EOS"));
+        EasyMock.replay(stateManager, recordCollector);
+
+        task = createStatefulTask(createConfig(EXACTLY_ONCE_V2, "100"), true);
+        task.initializeIfNeeded();
+        task.completeRestoration(noOpResetter -> { });
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Could you please use Mockito instead of EasyMock here? We are migrating to Mockito and we would like to have new tests using Mockito. See `TaskManagerTest` for an example where a test class uses Mockito and EasyMock.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 13]
```diff
@@ -255,6 +255,9 @@ public void completeRestoration(final java.util.function.Consumer<Set<TopicParti
                 resetOffsetsIfNeededAndInitializeMetadata(offsetResetter);
                 initializeTopology();
                 processorContext.initialize();
+                if (!eosEnabled) {
+                    maybeCheckpoint(true); // enforce checkpoint upon completing restoration
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Could you please remove the inline comment? I do not think it adds any information that is not already contained in the call to `maybeCheckpoint(true)`.

**Review by cadonna — COMMENTED** [review]
Thanks for the updates, @philipnee !

Here my feedback!

Sorry, that I missed that new tests should use Mockito. I could have told you before you started to add the tests.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 13 above)
Comment (philipnee) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: agreed, thanks for catching that.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java (hunk 5 above)
Comment (philipnee) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Ah, thanks.  I think it's from auto formatting from the IDE.  I'll revert these indentation changes.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java (hunk 7 above)
Comment (philipnee) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Sorry! I think I'm in the habit of using 8 spaces for "continuous indentation".

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java (hunk 9 above)
Comment (philipnee) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Thanks, i'm adopting your suggestion here.

**Comment by philipnee** [comment]
Hey @cadonna - Again, thanks for the reviews; sorry for the auto-cleanup for my IDE, I've cleaned up most of the unnecessary changes.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 14]
```diff
@@ -44,6 +44,7 @@
 import org.apache.kafka.common.utils.MockTime;
 import org.apache.kafka.common.utils.Utils;
 import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.TopologyConfig;
```
Comment (philipnee) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:47]: The IDE moved this, but it seems to be in an alphabetical sequence now. Can we keep it here?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 15]
```diff
@@ -2484,6 +2511,48 @@ public void shouldUpdateOffsetIfCorruptedRecordFollowsValid() {
...
+        task = createStatefulTask(createConfig(EXACTLY_ONCE_V2, "100"), true, processorStateManager);
+        task.initializeIfNeeded();
+        task.completeRestoration(noOpResetter -> { });
+        verify(processorStateManager, never()).checkpoint();
+        verify(processorStateManager, never()).changelogOffsets();
+        verify(recordCollector, never()).offsets();
+    }
+
+    private ProcessorStateManager mockStateManager() {
+        final ProcessorStateManager manager = spy(new ProcessorStateManager(
```
Comment (philipnee) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: I'm doing a partial mock here, to make sure the correct methods are being invoked.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java (hunk 14 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:47]: That is fine!

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java (hunk 15 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: I am afraid I cannot follow here. With Mockito you can define strict stubbings. That means Mockito will throw an exception if you have defined stubs that are not used. To use strict mocks, we usually specify 
```java
@MockitoSettings(strictness = Strictness.STRICT_STUBS)
```
on the test class level. Here, this does not work because I think (I am not sure) we use
```java
@RunWith(EasyMockRunner.class)
```    
What works is using
```java
    @Before
    public void setup() {
        mockito = Mockito.mockitoSession()
            .initMocks(this)
            .strictness(Strictness.STRICT_STUBS)
            .startMocking();
...
```
as first call in the setup method and 
```java
    @After
    public void cleanup() throws IOException {
        ...
        mockito.finishMocking();
    }
```
as last call in the teardown method.

With this the unit test becomes:
```java
    @Test
    public void shouldNotCheckpointAfterRestorationWhenExactlyOnceEnabled() {
        final ProcessorStateManager processorStateManager = mock(ProcessorStateManager.class);
        when(processorStateManager.taskType()).thenReturn(TaskType.ACTIVE);
        when(processorStateManager.taskId()).thenReturn(taskId);
        recordCollector = mock(RecordCollectorImpl.class);

        task = createStatefulTask(createConfig(EXACTLY_ONCE_V2, "100"), true, processorStateManager);
        task.initializeIfNeeded();
        task.completeRestoration(noOpResetter -> { });
        verify(processorStateManager, never()).checkpoint();
        verify(processorStateManager, never()).changelogOffsets();
        verify(recordCollector, never()).offsets();
    }
```

If you like you could extract this part into a method and reuse it in both unit tests (that is basically what you originally did, but w\o spy):
```java
        final ProcessorStateManager processorStateManager = mock(ProcessorStateManager.class);
        when(processorStateManager.taskType()).thenReturn(TaskType.ACTIVE);
        when(processorStateManager.taskId()).thenReturn(taskId);
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 16]
```diff
@@ -2484,6 +2511,48 @@ public void shouldUpdateOffsetIfCorruptedRecordFollowsValid() {
...
+    public void shouldCheckpointAfterRestorationWhenAtLeastOnceEnabled() {
+        final ProcessorStateManager processorStateManager = mockStateManager();
+        recordCollector = mock(RecordCollectorImpl.class);
+        doReturn(singletonMap(partition1, 1L)).when(recordCollector).offsets();
+
+        task = createStatefulTask(createConfig(AT_LEAST_ONCE, "100"), true, processorStateManager);
+        task.initializeIfNeeded();
+        task.completeRestoration(noOpResetter -> { });
+        verify(processorStateManager, times(1)).checkpoint();
+        verify(recordCollector, times(1)).offsets();
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: `times(1)` is default and can be omitted.

Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: `verify(recordCollector, times(1)).offsets();` is not strictly needed. The important thing in this test is that the checkpoint is written.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 17]
```diff
@@ -2484,6 +2511,48 @@ public void shouldUpdateOffsetIfCorruptedRecordFollowsValid() {
         );
     }
 
+    @Test
+    public void shouldCheckpointAfterRestorationWhenAtLeastOnceEnabled() {
+        final ProcessorStateManager processorStateManager = mockStateManager();
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:2522]:
```suggestion
        final ProcessorStateManager processorStateManager = mock(ProcessorStateManager.class);
        when(processorStateManager.taskType()).thenReturn(TaskType.ACTIVE);
        when(processorStateManager.taskId()).thenReturn(taskId);
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 18]
```diff
@@ -2484,6 +2511,48 @@ public void shouldUpdateOffsetIfCorruptedRecordFollowsValid() {
...
+        recordCollector = mock(RecordCollectorImpl.class);
+        doReturn(singletonMap(partition1, 1L)).when(recordCollector).offsets();
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: This is not needed since the checkpoint is forced.

**Review by cadonna — COMMENTED** [review]
Thanks for the updates, @philipnee !

I had just a couple of comments about the usage of the mocks and the unit tests.

**Comment by philipnee** [comment]
Thank you @cadonna - Based on your suggestions I configured the mockito strictness based on your suggestion.  I also cleaned the tests.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 19]
```diff
@@ -274,6 +287,7 @@ public void cleanup() throws IOException {
             task.closeDirty();
             task = null;
         }
+        mockito.finishMocking();
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Could you put this after `Utils.delete(BASE_DIR)` to ensure that the directory is deleted also if mockito throws an exception when a stub is unnecessary?

**Review by cadonna — COMMENTED** [review]
@philipnee  Thanks for the updates!

I have only one comment! After that is resolved, I think the PR is ready for merge.

**Comment by philipnee** [comment]
Hey @cadonna - sorry for the back and forth, your comment is addressed.

**Comment by cadonna** [comment]
Build failures are unrelated:
```
Build / JDK 8 and Scala 2.12 / kafka.security.authorizer.AuthorizerTest.testAllowAccessWithCustomPrincipal(String).quorum=
Build / JDK 11 and Scala 2.13 / kafka.security.authorizer.AuthorizerTest.testSuperUserHasAccess(String).quorum=kraft
Build / JDK 11 and Scala 2.13 / kafka.security.authorizer.AuthorizerTest.testAccessAllowedIfAllowAclExistsOnPrefixedResource(String).quorum=kraft
Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.processor.internals.StreamsAssignmentScaleTest.testFallbackPriorTaskAssignorLargePartitionCount
```

**Review by cadonna — APPROVED** [review]
@philipnee Thanks for the PR and the updates!

LGTM!

**Comment by mjsax** [comment]
@cadonna -- Should we cherry-pick to 3.4 branch?
